src/ring_theory/algebra.lean

```diff
@@ -576,10 +576,38 @@ def module.restrict_scalars : module R E :=
   add_smul  := by simp [add_smul],
   zero_smul := by simp [zero_smul] }
 
+/--
+The identity function, considered as an `R`-linear map from an `R`-algebra `S` to itself,
+with `module.restrict_scalars R S S` as the module structure in the source,
+and `algebra.to_module` as the module structure in the target.
```

Comment:
Is there no reasonable way to change one of the definitions to have defeqness?

Reply:
I'm pretty sure it's not. The problem is that the `restrict_scalars` scalar multiplication uses the `ring_hom` field of the algebra. By design there is a field of `algebra` asserting that this is equal to the usual scalar multiplication, but there's way to demand this field is `rfl`.

If @urkud is back from exams he could give a second opinion.

Reply:
The alternative would be to decide that `module.restrict_scalars` is evil, as it puts a second module structure on `E`. Instead, we could define a type synonym, `restrict_scalars R E := E`, and then provide an instance `module R (restrict_scalars R E)`, given `algebra R S` and `module S E`.

This is perhaps a more dramatic change, but perhaps better in the long run. I'd like advice from others before proceeding, however.

Reply:
I see that this would be pretty painful, for example requiring us to be much more verbose whenever we treat a complex vector space as a real vector space.

Perhaps there's also the compromise solution: keep the current def, but also add the type synonym, which can be equipped with the instance.

Reply:
I am not really convinced by the type synonym: having two module structures with different rings on the same type is definitely not evil, and it is precisely to make this possible that the ring is explicit in the module instance. If having a type synonym makes things smoother in your use case, though, then ok, but I would like to be sure you have a real use case.

Reply:
@sgouezel, I think I really do want to have the type synonym.

1. To make more explicit what is going on in the code.
2. Because without it `module.restrict_scalars` is a "bad" instance: asking for `module R E` requires typeclass search to go off and find some type `S`, with `algebra R S` and `module S E`.